lib/mongoid/persistence_context.rb

```diff
@@ -232,7 +231,35 @@ def clear(object, cluster = nil, original_context = nil)
           end
         end
       ensure
-        Thread.current["[mongoid][#{object.object_id}]:context"] = original_context
+        store_context(object, original_context)
+      end
+
+      private
+
+      # Get the persistence context for a given object from the thread local
+      #   storage.
+      #
+      # @param [ Object ] object Object to get the persistance context for.
+      #
+      # @return [ Mongoid::PersistenceContext | nil ] The persistence context
+      #   for the object if previously stored, otherwise nil.
+      #
+      # @api private
+      def get_context(object)
+        Thread.current[:context] ||= {}
+        Thread.current[:context][object.object_id]
+      end
+
+      # Store persistence context for a given object in the thread local
+      #   storage.
+      #
+      # @param [ Object ] object Object to store the persistance context for.
+      # @param [ Mongoid::PersistenceContext ] context Context to store
+      #
+      # @api private
+      def store_context(object, context)
+        Thread.current[:context] ||= {}
+        Thread.current[:context][object.object_id] = context
```

Comment:
It's not clear to me how this fixes the memory leak, since the context hash itself is never reset. It will keep growing and growing as more objects accumulate.

Reply:
The problem was that a new symbol was created every time we store the context, and never released. If we use `object_id` as a key, no new object is created.
However, you are right that we can do better here.